test/e2e/vmagent_test.go

```diff
@@ -289,5 +300,378 @@ func createExternalNodeCRD(data *TestData, nodeName string, ifName string, ip st
 	return data.crdClient.CrdV1alpha1().ExternalNodes(namespace).Create(context.TODO(), testEn.Get(), metav1.CreateOptions{})
 }
+
+func testExternalNodeWithANP(t *testing.T, data *TestData, vmList []vmInfo) {
+	if len(vmList) < 2 {
+		t.Skipf("Skipping test as it requires 2 different VMs but the setup has %d", len(vmList))
+	}
+	t.Run("testANPOnLinuxVM", func(t *testing.T) { testANPOnVMs(t, data, vmList, linuxOS) })
+	t.Run("testANPOnWindowsVM", func(t *testing.T) { testANPOnVMs(t, data, vmList, windowsOS) })
+}
+
+func testANPOnVMs(t *testing.T, data *TestData, vmList []vmInfo, osType string) {
+	appliedToVM, peerVM, err := getVMsByOSType(vmList, osType)
+	if err != nil {
+		t.Skipf("Skip case testANPOnVMs: %v", err)
+	}
+	// Test TCP rules in ANP
+	t.Run("testANPOnExternalNodeWithTCP", func(t *testing.T) {
+		// Use ExternalEntity in an ingress rule configuration.
+		testANPProtocolTCPOrUDP(t, data, "anp-vmagent-ingress-tcp-entity", namespace, *appliedToVM, peerVM, ProtocolTCP, true, crdv1alpha1.RuleActionDrop, true)
+		// Use IP in an egress rule configuration.
+		testANPProtocolTCPOrUDP(t, data, "anp-vmagent-egress-tcp-ip", namespace, *appliedToVM, peerVM, ProtocolTCP, false, crdv1alpha1.RuleActionDrop, false)
+	})
+	// Test UDP rules in ANP
+	t.Run("testANPOnExternalNodeWithUDP", func(t *testing.T) {
+		testANPProtocolTCPOrUDP(t, data, "anp-vmagent-ingress-udp-entity", namespace, *appliedToVM, peerVM, ProtocolUDP, true, crdv1alpha1.RuleActionReject, false)
+	})
+	// Test ICMP rules in ANP
+	t.Run("testANPOnExternalNodeWithICMP", func(t *testing.T) {
+		testANPProtocolICMP(t, data, "anp-vmagent-ingress-icmp-ip", namespace, *appliedToVM, crdv1alpha1.RuleActionDrop)
+	})
+	// Test FQDN rules in ANP
+	t.Run("testANPOnExternalNodeWithFQDN", func(t *testing.T) {
+		testANPWithFQDN(t, data, "anp-vmagent-fqdn", namespace, *appliedToVM, []string{"www.facebook.com"}, []string{"docs.google.com"}, []string{"maps.google.com"})
+	})
+}
+
+// getVMsByOSType returns the appliedTo VM and a different VM to run test. The appliedTo VM is configured with the given
+// osType, and the other VM returned is the next one of the appliedTo VM in the given vmList.
+func getVMsByOSType(vmList []vmInfo, osType string) (*vmInfo, *vmInfo, error) {
+	for i := range vmList {
+		if vmList[i].osType == osType {
+			return &vmList[i], &vmList[(i+1)%len(vmList)], nil
+		}
+	}
+	return nil, nil, fmt.Errorf("not found a VM configured with OS type %s in vmList", osType)
+}
+
+func testANPWithFQDN(t *testing.T, data *TestData, name string, namespace string, appliedToVM vmInfo, allowedURLs []string, droppedURLs []string, rejectedURLs []string) {
+	var err error
+	allURLs := append(append(allowedURLs, droppedURLs...), rejectedURLs...)
+	for _, url := range allURLs {
+		err := runCurlCommandOnVM(data, appliedToVM, url, crdv1alpha1.RuleActionAllow)
+		assert.Nil(t, err, fmt.Sprintf("Failed to run curl command on URL %s on VM %s: %v", url, appliedToVM.nodeName, err))
```

Comment:
```suggestion
		assert.NoError(t, err, "Failed to run curl command on URL %s on VM %s", url, appliedToVM.nodeName)
```
1. assert,NoError supports formating message and arguments like Sprintf
2. NoError dumps the unexpected error

Reply:
updated.